# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [2]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [3]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [4]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [5]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

In [6]:
pd.crosstab(df["Gender"], df["SalesTraining"], margins=True, normalize=False)

SalesTraining    N    Y  All
Gender                      
F              212  277  489
M              201  309  510
All            413  586  999

In [7]:
pd.crosstab(df["Gender"], df["SalesTraining"], margins=True, normalize=True)

SalesTraining         N         Y       All
Gender                                     
F              0.212212  0.277277  0.489489
M              0.201201  0.309309  0.510511
All            0.413413  0.586587  1.000000

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [8]:
#Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows.
poke = pd.read_csv("pokemon.csv")

In [9]:
poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [10]:
#Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts(normalize=True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [11]:
#Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts(normalize=False)

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [12]:
#Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary).
pd.crosstab(poke["Stage"], poke["Legendary"], margins=True, normalize=True)

Legendary     False     True       All
Stage                                 
1          0.496689  0.02649  0.523179
2          0.370861  0.00000  0.370861
3          0.105960  0.00000  0.105960
All        0.973510  0.02649  1.000000

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [13]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [14]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [15]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
Name: Hours Worked, dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [16]:
poke["Attack"].mean()
#this is the average value of all the pokemon attack stats combined

72.54966887417218

In [17]:
poke["Defense"].median()
#The median is the value in the middle of a data set, meaning that 50% of data points have a value smaller or equal to the median and 50% of data points have a value higher or equal to the median.

65.0

In [18]:
poke["Speed"].mode()
#The mode is a statistical term that refers to the most frequently occurring number found in a set of numbers.

0    70
1    90
Name: Speed, dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [19]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [20]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [21]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [22]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [23]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [24]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [25]:
poke_range = poke["HP"].max()-poke["HP"].min()
print(poke_range)

240


In [26]:
print("Mean special for pokemon is:", poke["SpAtk"].mean())

poke["SpAtk"].std()

Mean special for pokemon is: 67.13907284768212


28.53419930191353

In [27]:
np.percentile(poke["Attack"], 80)
            

95.0

In [28]:
np.percentile(poke["Attack"], 15)

45.0

In [29]:
poke["SpDef"].std()

24.197925585657266

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [30]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [31]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [32]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_Salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_Salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [33]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_Salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [34]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_Salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [35]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [36]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [37]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [38]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [39]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [40]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [41]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [42]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [43]:
df2 = pd.read_excel("babies.xlsx")

df2

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
3     123        NaN       0  36.0    69.0   190.0    0.0
4     108      282.0       0  23.0    67.0   125.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 7 columns]

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [44]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


In [45]:
df2.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [46]:
df2.describe()

bwt    gestation       parity          age       height  \
count  1236.000000  1223.000000  1236.000000  1234.000000  1214.000000   
mean    119.576861   279.338512     0.254854    27.255267    64.047776   
std      18.236452    16.027693     0.435956     5.781405     2.533409   
min      55.000000   148.000000     0.000000    15.000000    53.000000   
25%     108.750000   272.000000     0.000000    23.000000    62.000000   
50%     120.000000   280.000000     0.000000    26.000000    64.000000   
75%     131.000000   288.000000     1.000000    31.000000    66.000000   
max     176.000000   353.000000     1.000000    45.000000    72.000000   

            weight        smoke  
count  1200.000000  1226.000000  
mean    131.411667     0.394780  
std      53.687343     0.489003  
min      27.000000     0.000000  
25%     114.000000     0.000000  
50%     125.000000     0.000000  
75%     140.000000     1.000000  
max    1500.000000     1.000000

In [47]:
df2.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [48]:
df2.dropna(inplace=True)

df2

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [49]:
#The two qualitative datasets in the dataframe 2 (df2) are  parity and smoke columns/categories. 

#parity - previously pregnant (0 = no; 1 = yes); smoke - smoking status of mother (0 = nonsmoker; 1 = smoker

#Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). 

#These data can only be classified into one group of categories. 

#qualitative variables are also called categorical variables. Qualitative data also includes text and string data.

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [50]:
df2

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [51]:
df2["parity"].replace([1,0],["Yes","No"], inplace = True)
df2

bwt  gestation parity   age  height  weight  smoke
0     120      284.0     No  27.0    62.0   100.0    0.0
1     113      282.0     No  33.0    64.0   135.0    0.0
2     128      279.0     No  28.0    64.0    77.0    1.0
4     108      282.0     No  23.0    67.0   125.0    1.0
5     136      286.0     No  25.0    62.0    93.0    0.0
...   ...        ...    ...   ...     ...     ...    ...
1231  113      275.0    Yes  27.0    60.0   100.0    0.0
1232  128      265.0     No  24.0    67.0   120.0    0.0
1233  130      291.0     No  30.0    65.0    67.0    1.0
1234  125      281.0    Yes  21.0    65.0   110.0    0.0
1235  117      297.0     No  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [52]:
df2["smoke"].replace([1,0],["Yes","No"], inplace=True)
df2

bwt  gestation parity   age  height  weight smoke
0     120      284.0     No  27.0    62.0   100.0    No
1     113      282.0     No  33.0    64.0   135.0    No
2     128      279.0     No  28.0    64.0    77.0   Yes
4     108      282.0     No  23.0    67.0   125.0   Yes
5     136      286.0     No  25.0    62.0    93.0    No
...   ...        ...    ...   ...     ...     ...   ...
1231  113      275.0    Yes  27.0    60.0   100.0    No
1232  128      265.0     No  24.0    67.0   120.0    No
1233  130      291.0     No  30.0    65.0    67.0   Yes
1234  125      281.0    Yes  21.0    65.0   110.0    No
1235  117      297.0     No  38.0    65.0   129.0    No

[1174 rows x 7 columns]

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [53]:
#bwt - birth weight of newborn baby
#gestation - gestation length (weeks)
#age - age of mother
#height - height of mother (inches)
#weight - weight of mother (pounds)


In [54]:
df2.describe()

bwt    gestation          age       height       weight
count  1174.000000  1174.000000  1174.000000  1174.000000  1174.000000
mean    119.462521   279.101363    27.228279    64.049404   131.326235
std      18.328671    16.010305     5.817839     2.526102    54.103828
min      55.000000   148.000000    15.000000    53.000000    27.000000
25%     108.000000   272.000000    23.000000    62.000000   114.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   139.750000
max     176.000000   353.000000    45.000000    72.000000  1500.000000

In [55]:
#Q1 (Quartile 1)
df2["bwt"].quantile(.25)

108.0

In [56]:
#Q1 (Quartile 2)
df2["bwt"].quantile(.50)

120.0

In [57]:
#Q1 (Quartile 3)
df2["bwt"].quantile(.75)

131.0

In [58]:
dfq = df2.copy() ## Create a copy of original dataset

df2.shape #checking the original shape of the df2 dataframe/dataset

(1174, 7)

In [59]:
## calculate the quartiles

q1 = dfq["bwt"].quantile(.25)
q3 = dfq["bwt"].quantile(.75)

print("Q1:" , q1)
print("Q3:" , q3)

Q1: 108.0
Q3: 131.0


In [60]:
iqr = q3 - q1

print("IQR:",iqr)

IQR: 23.0


In [61]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 165.5
Lower Limit: 73.5


In [62]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['bwt'] > top) | (dfq['bwt'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169],
           dtype='int64')


In [63]:
## what values can we find at these index locations?

dfq.iloc[[239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169]]

bwt  gestation parity   age  height  weight smoke
261    81      256.0     No  30.0    64.0   148.0   Yes
332   103      273.0    Yes  22.0    64.0   110.0   Yes
388    98      275.0     No  25.0    65.0   112.0   Yes
497   101      289.0    Yes  31.0    60.0   125.0    No
537   147      277.0     No  30.0    68.0   160.0    No
566   133      280.0    Yes  25.0    61.0   130.0    No
593   118      297.0     No  35.0    68.0   140.0   Yes
632   176      293.0    Yes  19.0    68.0   180.0    No
676   100      275.0     No  26.0    60.0   115.0    No
757   155      279.0     No  33.0    61.0   125.0    No
787   125      272.0     No  30.0    60.0    96.0    No
796   152      286.0    Yes  19.0    67.0   135.0    No
881   128      272.0    Yes  18.0    67.0   109.0    No
958   133      290.0     No  21.0    64.0   145.0    No
967   119      273.0     No  35.0    65.0   125.0   Yes
1035   65      237.0     No  31.0    67.0   130.0    No
1079  119      286.0    Yes  33.0    67.0   137.0    No
1093  108      256.0    Yes  26.0    67.0   130.0    No
1121  128      277.0     No  39.0    61.0   120.0    No
1123  133      289.0     No  22.0    65.0   123.0   Yes
1157  141      293.0     No  28.0    61.0   125.0    No
1200   97      255.0    Yes  22.0    63.0   107.0   Yes
1209  144      283.0    Yes  25.0    66.0   140.0    No
1231  113      275.0    Yes  27.0    60.0   100.0    No

In [64]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)
## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(1150, 7)


In [65]:
df2.describe()

bwt    gestation          age       height       weight
count  1174.000000  1174.000000  1174.000000  1174.000000  1174.000000
mean    119.462521   279.101363    27.228279    64.049404   131.326235
std      18.328671    16.010305     5.817839     2.526102    54.103828
min      55.000000   148.000000    15.000000    53.000000    27.000000
25%     108.000000   272.000000    23.000000    62.000000   114.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   139.750000
max     176.000000   353.000000    45.000000    72.000000  1500.000000

In [66]:
## Calculate quartiles
q1 = dfq["gestation"].quantile(.25)
q3 = dfq["gestation"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 272.0
Q3: 288.0


In [67]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 16.0


In [68]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 312.0
Lower Limit: 248.0


In [69]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['gestation'] > top) | (dfq['gestation'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  373,
             394,  440,  460,  484,  511,  523,  630,  685,  699,  710,  726,
             746,  761,  769,  778,  784,  828,  830,  833,  869,  927,  952,
             969, 1002, 1020, 1026, 1074, 1134, 1140, 1142, 1146, 1152, 1172,
            1178, 1199, 1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [70]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(1088, 7)


In [71]:
## Calculate quartiles
q1 = dfq["age"].quantile(.25)
q3 = dfq["age"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 23.0
Q3: 31.0


In [72]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 8.0


In [73]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 43.0
Lower Limit: 11.0


In [74]:
iqr_outliers = dfq.loc[(dfq['age'] > top) | (dfq['age'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([633, 1070], dtype='int64')


In [75]:
dfq.iloc[[633, 1070]]

bwt  gestation parity   age  height  weight smoke
721   117      298.0    Yes  22.0    64.0   160.0    No
1214  118      293.0     No  21.0    63.0   103.0    No

In [76]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(1086, 7)


In [77]:
dfq.describe()

bwt    gestation          age       height       weight
count  1086.000000  1086.000000  1086.000000  1086.000000  1086.000000
mean    120.254144   279.902394    27.233886    64.050645   131.183241
std      16.469938    11.411679     5.725084     2.537137    55.871516
min      75.000000   248.000000    15.000000    53.000000    27.000000
25%     110.000000   273.000000    23.000000    62.000000   114.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   137.000000
max     165.000000   312.000000    43.000000    72.000000  1500.000000

In [78]:
## Calculate quartiles
q1 = dfq["height"].quantile(.25)
q3 = dfq["height"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 62.0
Q3: 66.0


In [79]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 4.0


In [80]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 72.0
Lower Limit: 56.0


In [81]:
iqr_outliers = dfq.loc[(dfq['height'] > top) | (dfq['height'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([434, 1208], dtype='int64')


In [82]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(1084, 7)


In [83]:
dfq.describe()

bwt    gestation          age       height       weight
count  1084.000000  1084.000000  1084.000000  1084.000000  1084.000000
mean    120.211255   279.916974    27.221402    64.070111   131.179889
std      16.454437    11.410599     5.718937     2.498508    55.914292
min      75.000000   248.000000    15.000000    56.000000    27.000000
25%     110.000000   273.000000    23.000000    62.000000   114.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   137.000000
max     165.000000   312.000000    43.000000    72.000000  1500.000000

In [84]:
## Calculate quartiles
q1 = dfq["weight"].quantile(.25)
q3 = dfq["weight"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 114.0
Q3: 137.0


In [85]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 23.0


In [86]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 171.5
Lower Limit: 79.5


In [87]:
iqr_outliers = dfq.loc[(dfq['weight'] > top) | (dfq['weight'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   2,   14,   21,   23,   27,   41,   88,  117,  125,  147,  149,
             162,  180,  181,  183,  222,  287,  310,  314,  344,  411,  426,
             458,  509,  512,  522,  528,  563,  595,  608,  622,  695,  696,
             723,  733,  849,  858,  859,  865,  888,  911,  924,  935, 1008,
            1154, 1167, 1212, 1233],
           dtype='int64')


In [88]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(1036, 7)


7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [94]:
dfq

bwt  gestation parity   age  height  weight smoke
0     120      284.0     No  27.0    62.0   100.0    No
1     113      282.0     No  33.0    64.0   135.0    No
4     108      282.0     No  23.0    67.0   125.0   Yes
5     136      286.0     No  25.0    62.0    93.0    No
8     120      289.0     No  25.0    62.0   125.0    No
...   ...        ...    ...   ...     ...     ...   ...
1230  132      270.0     No  27.0    65.0   126.0    No
1231  113      275.0    Yes  27.0    60.0   100.0    No
1232  128      265.0     No  24.0    67.0   120.0    No
1234  125      281.0    Yes  21.0    65.0   110.0    No
1235  117      297.0     No  38.0    65.0   129.0    No

[1036 rows x 7 columns]

Qualitative vs Quantitative Data
Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called numeric variables.

Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called categorical variables. Qualitative data also includes text and string data.

In [95]:
dfq["parity"].value_counts()

No     754
Yes    282
Name: parity, dtype: int64

In [96]:
dfq["parity"].value_counts(normalize=True)

No     0.727799
Yes    0.272201
Name: parity, dtype: float64

In [97]:
dfq["smoke"].value_counts()

No     631
Yes    405
Name: smoke, dtype: int64

In [101]:
dfq["smoke"].value_counts(normalize=True)

No     0.609073
Yes    0.390927
Name: smoke, dtype: float64

In [103]:
pd.crosstab(dfq["parity"],dfq["smoke"],margins=True,normalize=True)

smoke         No       Yes       All
parity                              
No      0.443050  0.284749  0.727799
Yes     0.166023  0.106178  0.272201
All     0.609073  0.390927  1.000000

In [104]:
pd.crosstab(dfq["parity"],dfq["smoke"],margins=True,normalize=False)

smoke    No  Yes   All
parity                
No      459  295   754
Yes     172  110   282
All     631  405  1036

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [105]:
dfq.describe()

bwt    gestation          age       height       weight
count  1036.000000  1036.000000  1036.000000  1036.000000  1036.000000
mean    120.175676   279.984556    27.159266    64.009653   125.656371
std      16.404372    11.396404     5.723814     2.481215    16.581046
min      75.000000   248.000000    15.000000    56.000000    81.000000
25%     109.750000   273.000000    23.000000    62.000000   113.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   135.000000
max     165.000000   312.000000    43.000000    72.000000   171.000000

In [107]:
dfq.shape

(1036, 7)

In [108]:
dfq["bwt"].mean()

120.17567567567568

In [109]:
dfq["bwt"].median()

120.0

In [110]:
dfq["bwt"].mode()

0    117
Name: bwt, dtype: int64

In [111]:
#for bwt(birth weight) it shows the values of mean,median and mode are evenly distributed since their values are not far off from each orther
#this is a normal bell curve (mean = median)

In [112]:
dfq["gestation"].mean()

279.984555984556

In [113]:
dfq["gestation"].median()

280.0

In [114]:
dfq["gestation"].mode()

0    282.0
Name: gestation, dtype: float64

In [ ]:
#for gestation, it shows that the value of the mean and median are almost exactly the same
#this represents a normal distribution (normal bell curve) mean = median

In [117]:
dfq["age"].mean()

27.159266409266408

In [118]:
dfq["age"].median()

26.0

In [119]:
dfq["age"].mode()

0    24.0
Name: age, dtype: float64

In [ ]:
#the age column is still a normal distribution (normal bell curve) base on the values of the mean, median and mode but slightly slanted

In [120]:
dfq["height"].mean()

64.00965250965251

In [121]:
dfq["height"].median()

64.0

In [122]:
dfq["height"].mode()

0    65.0
Name: height, dtype: float64

In [123]:
#The height variable(column) is also a normal distribution based on the mean, median and mode.

In [124]:
dfq["weight"].mean()

125.65637065637065

In [125]:
dfq["weight"].median()

125.0

In [126]:
dfq["weight"].mode()

0    130.0
Name: weight, dtype: float64

In [127]:
#The height variable(column) is also a normal distribution based on the mean, median and mode.

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [128]:
dfq.describe()

bwt    gestation          age       height       weight
count  1036.000000  1036.000000  1036.000000  1036.000000  1036.000000
mean    120.175676   279.984556    27.159266    64.009653   125.656371
std      16.404372    11.396404     5.723814     2.481215    16.581046
min      75.000000   248.000000    15.000000    56.000000    81.000000
25%     109.750000   273.000000    23.000000    62.000000   113.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   135.000000
max     165.000000   312.000000    43.000000    72.000000   171.000000

In [133]:
bwt_range = dfq["bwt"].max()-dfq["bwt"].min()

print(bwt_range)

90


In [134]:
bwt_range = dfq["gestation"].max()-dfq["gestation"].min()

print(bwt_range)

64.0


In [135]:
bwt_range = dfq["age"].max()-dfq["age"].min()

print(bwt_range)

28.0


In [136]:
bwt_range = dfq["height"].max()-dfq["height"].min()

print(bwt_range)

16.0


In [137]:
bwt_range = dfq["weight"].max()-dfq["weight"].min()

print(bwt_range)

90.0


In [138]:
#according to the statistical outlook of ranges per quantitative column, bwt and weight had the largest value. 
#this could mean that new born babies data is highly devirsified in terms of birthweight and actual weight

#A low range in the height of base on the dataset/dataframe of babies indicate that the group is all very similar in terms of height.

In [139]:
#Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

dfq.describe()

bwt    gestation          age       height       weight
count  1036.000000  1036.000000  1036.000000  1036.000000  1036.000000
mean    120.175676   279.984556    27.159266    64.009653   125.656371
std      16.404372    11.396404     5.723814     2.481215    16.581046
min      75.000000   248.000000    15.000000    56.000000    81.000000
25%     109.750000   273.000000    23.000000    62.000000   113.000000
50%     120.000000   280.000000    26.000000    64.000000   125.000000
75%     131.000000   288.000000    31.000000    66.000000   135.000000
max     165.000000   312.000000    43.000000    72.000000   171.000000

In [141]:
dfq.shape

(1036, 7)

In [ ]:
#base on the outlook statistics with dfq.describe(); the SMALLEST std per variable is "height" and the HIGHEST std "weight"

#this means that the "height" more tightly clustered around a central value or more condensed
#the "weight" std signify that the values are more SPREAD within the dataframe